rfc/20250711-Middleware-For-Enhanced-Operations.md

```diff
@@ -0,0 +1,378 @@
+- **Simplified workflows**: Middleware runs as part of normal OpenTofu operations
+- **Bidirectional communication**: Middleware can fail operations and store metadata in state
+- **Language agnostic**: Write middleware in any language
+
+By offering this extensibility through a well-defined protocol, we aim to enable organizations to bring their existing policy and governance tools into the OpenTofu execution flow while fostering innovation in new use cases that weren't possible with external analysis alone.
+
+The architecture of this middleware system is inspired by the recent Model Context Protocol (MCP) approach, which has proven successful for extending AI assistants through local server processes communicating via JSON-RPC over stdio (along with other transports). This design pattern offers simplicity, language independence, and process isolation - qualities that align perfectly with OpenTofu's needs for a stable, extensible middleware system without the overhead of producing a full SDK or requiring middleware authors to be locked into a specific languge/ecosystem. You can find more information about this specification [at modelcontextprotocol.io/docs](https://modelcontextprotocol.io/docs/concepts/architecture).
+
+## Approach
+To provide users with a flexible and powerful way to extend OpenTofu's functionality, I propose implementing a middleware system that communicates with external processes via JSON-RPC 2.0 over STDIO. The middleware will be invoked at specific hook points during OpenTofu operations, allowing users to inspect, validate, and augment the behavior of resource operations without requiring changes to OpenTofu's core.
+
+### Hook Points for Operations
+The middleware system will provide hook points at two levels:
+
+**Resource-level hooks** (called for each individual resource/data source):
+- **pre-plan**
+- **post-plan**
+- **pre-apply**
+- **post-apply**
+- **pre-refresh**
+- **post-refresh**
+
+**Operation-level hooks** (called once per operation):
+- **init-stage-start**
+- **init-stage-complete**
+- **plan-stage-start**
+- **plan-stage-complete**
+- **apply-stage-start**
+- **apply-stage-complete**
+
+Additional hooks can be added in future versions as use cases emerge. It is possible that we should also consider hooks that handle failures/errors too.
+
+### Middleware Attachments
+
+For now I propose that middleware should be "attachable" to the entire project, or to a specific provider instance. This opens up the prospect of having provider-extension middleware or project-wide enforcement of rules.
+
+### Configuration Design
+Users will configure middleware through HCL blocks in their OpenTofu configuration:
+
+```hcl
+
+middleware "naming_convention_checker" {
+  command = "lint-opentofu-names"
+}
+
+middleware "cost_estimator" {
+  command = "python3"
+  args    = ["cost_estimator.py"]
+  env = {
+    API_KEY = var.cost_api_key
+  }
```

Comment:
Im not sure this is going to be needed, they could just have the ENV be part of the environment thats running tofu. We'd be effectively setting the environment twice here, right? If folks need to consume variables in their runtime, they could just export those variables before running tofu. I think passing variables through the `args` parameter would be a more valid use-case.

Reply:
>TBH, I am concerned about the possible plethora of malicious implementations that could leak sensitive user information, like tokens and other sensitive bits.

IDK my thought is this is no different than a provider which can pick up the runtimes environment. Like the AWS provider for example can pick up `AWS_ACCESS_KEY_ID` etc etc.

Maybe we do both? Just like providers? You can explicitly pass things in but you can also just let the middleware pick up things from the current environment.